substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1156,17 +1150,21 @@ impl<T: Config> Pallet<T> {
         };
 
         // Handle grace
-        let contract = Self::handle_grace(&mut contract, usable_balance, amount_due)?;
+        let contract = Self::handle_grace(
+            &mut contract,
+            usable_balance,
+            contract_lock.amount_locked + amount_due,
+        )?;
+
+        // Handle contract lock operations
+        Self::handle_lock(contract, amount_due)?;
```

Comment:
If we put it here I am not sure we avoid trying to distribute rewards while in grace period